### DISCUSSION TITLE
How Can I Access `localstorage` from server Component?

### DISCUSSION THREAD

**Discussion by clevermiraz** [vercel/next.js#58434] [App Router]
Is There Any Way to Get `localstorage` directly In the server component?

Or Now I Think In the Server Component,
make a client Component then Get The `localstorage` from there then Use It.

Basically, In `localstorage` we store `accessToken` we need that For Fetching Private Data.

**Comment by icyJoseph** [comment]
Well, the short ~lazy~ answer, is that you can't, because Server Components exist only on the Server Side. They do not have access to browser APIs, let alone data stored within those APIs. 

It is the equivalent is trying to access the window innerWidth from a Server Component.

That being said, Server Actions, are the alternative, to activate server side work, from the client, and use the result of calling that Server Action to populate the UI. It'd be the equivalent of doing `useState + useEffect`. 

Otherwise, you'd need to re-think your approach. It seems like you'd benefit from using httpOnly cookies, for this use case.

**Reply by clevermiraz** [reply]
Okay, I Understand.

So Now I Think, getting tokens from cookies is a good idea.

Or Should I Need to Rethink?

for getting cookies I'm using `next/headers`

https://nextjs.org/docs/app/api-reference/functions/headers

**Comment by prkagrawal** [comment]
First of all, private data should never be stored in local storage, [why not](https://www.rdegges.com/2018/please-stop-using-local-storage/)

Now you can use http-only  cookie or auth header for storing user token. Both have their [pros and cons](https://stackoverflow.com/questions/17000835/token-authentication-vs-cookies)

Also if you are using http-only cookie you will have to use [https for development](https://web.dev/articles/when-to-use-local-https), because they can't be tested on localhost.

**Comment by Shyam-Raghuwanshi** [comment]
Hi @prkagrawal @clevermiraz, i have this code and i want a value from localStorage and the value is not a private data the value is for redirect the user when user come on the website , and this is not a api call from anywhere so i can't set the headers

```
export async function getServerSideProps({ req, res }: GetServerSidePropsContext) {
  const session = await getServerSession({ req, res });

  if (!session?.user?.id) {
    return { redirect: { permanent: false, destination: "/auth/login" } };
  }

  return { redirect: { permanent: false, destination: "/event-types" } };
}

export default RedirectPage;
```
then what should i do?

**Comment by hb20007** [comment]
`localStorage` exists only on the client. The correct way to access it is to use a client component. And since client components also execute on the server, you need to disable SSR on that specific component to make sure it only executes on the client. You can find an example in [this SO answer](https://stackoverflow.com/a/79755525/4304516).

**Comment by LienThuan04** [comment]
No — localStorage is browser-only. Server Components run on the server where window doesn't exist, so it will always be undefined.

The correct fix: store accessToken in a cookie instead of localStorage.

Cookies are sent with every request, so Server Components can read them directly.

Step 1 — Set the cookie on login:
import { cookies } from "next/headers";

// in your login Server Action or API route
cookies().set("accessToken", token, {
  httpOnly: true,
  secure: true,
  sameSite: "strict",
  path: "/",
});

Step 2 — Read it in any Server Component:
import { cookies } from "next/headers";

export default async function Page() {
  const token = cookies().get("accessToken")?.value;

  const res = await fetch("https://your-api.com/private", {
    headers: { Authorization: `Bearer ${token}` },
  });

  const data = await res.json();
}

Why not the Client Component workaround:
- Client Component → reads localStorage → passes up to Server Component: not possible, data only flows down in React
- You'd end up fetching on the client anyway, losing all SSR benefits

Bonus — localStorage vs httpOnly cookie for tokens:

┌──────────────────────┬──────────────┬─────────────────┐
│                      │ localStorage │ httpOnly cookie │
├──────────────────────┼──────────────┼─────────────────┤
│ Accessible on server │ No           │ Yes             │
├──────────────────────┼──────────────┼─────────────────┤
│ XSS risk             │ High         │ None            │
├──────────────────────┼──────────────┼─────────────────┤
│ Works with SSR       │ No           │ Yes             │
└──────────────────────┴──────────────┴─────────────────┘

**Comment by xdatabox** [comment]
No, you cannot access localStorage directly from a Server Component.

Server Components run on the server (Node.js environment), so browser-specific APIs like localStorage, window, or document are not available.

If you need to use localStorage, you must move that logic into a Client Component by adding "use client" at the top.

Example:

"use client";

import { useEffect } from "react";

export default function Component() {
  useEffect(() => {
    const token = localStorage.getItem("accessToken");
    console.log(token);
  }, []);

  return <div>Client Component</div>;
}

**Comment by vinay-sh87** [comment]
If the redirect decision depends on a value stored in `localStorage`, then `getServerSideProps` (or a Server Component) cannot perform that redirect because the server has no access to browser storage.

You have three options:

1. Store the value in a cookie instead, so the server can read it and redirect before rendering.

2. Perform the redirect on the client:

```tsx
"use client";

import { useEffect } from "react";
import { useRouter } from "next/navigation";

export default function RedirectPage() {
  const router = useRouter();

  useEffect(() => {
    const destination = localStorage.getItem("redirectTo");

    if (destination) {
      router.replace(destination);
    } else {
      router.replace("/event-types");
    }
  }, [router]);

  return null;
}
```

3. Persist the value to the server (database/session/cookie) before the page load if the redirect must happen during SSR.

As a rule of thumb: if the server needs the value, don't store it exclusively in `localStorage`.

**Comment by Ishant1302** [comment]
> Is There Any Way to Get `localstorage` directly In the server component?
> 
> Or Now I Think In the Server Component, make a client Component then Get The `localstorage` from there then Use It.
> 
> Basically, In `localstorage` we store `accessToken` we need that For Fetching Private Data.

No, Server Components cannot access localStorage directly. localStorage only exists in the browser, while Server Components run on the server.

If you need an access token in a Server Component, storing it in localStorage is usually not the best approach. The recommended pattern is to store authentication data in cookies (preferably HTTP-only cookies), since cookies are available to the server during rendering.

While you can create a Client Component, read localStorage, and pass the value around on the client, that token still won't be available during the initial Server Component render. If the server needs the token to fetch private data, move the token to cookies instead of localStorage.

**Reply by LienThuan04** [reply]
Good question — it seems contradictory, but the two cases are actually opposite scenarios. The deciding factor isn't "sync vs async," it's **how long the operation blocks the event loop**.

**bcrypt:** intentionally slow. Cost factor 10 = ~100ms of pure CPU grinding by design (that's the whole point of a password hash). Blocking the event loop for 100ms per request is catastrophic under load.

**better-sqlite3:** local SQLite queries are *microseconds to low-milliseconds*. SQLite is an embedded database — no network, no socket, just a memory/disk read on the same machine. The "I/O" is so fast that the overhead of going async (callback scheduling, thread pool handoff, serialization) costs *more* than just doing the work synchronously.

**Why `better-sqlite3` sync actually wins:**

The `sqlite3` async API hands each query to libuv's thread pool. For an operation that takes 50 microseconds, the thread pool handoff + context switch + callback queueing dominates the actual work. You pay coordination overhead for nothing.

```
sqlite3 (async):       [query 50µs] + [thread handoff ~100µs+] = overhead dominates
better-sqlite3 (sync): [query 50µs] = done, event loop barely notices
```

Benchmarks consistently show `better-sqlite3` is several times faster for typical workloads precisely because it skips that machinery.

**The unifying principle:**

Don't ask "is it sync or async?" Ask **"how long will this block the single thread, and is that duration acceptable?"**

| Operation | Duration | Block the event loop? |
|---|---|---|
| bcrypt hash (cost 10) | ~100ms | ❌ No — offload (async) |
| SQLite local query | ~50µs–1ms | ✅ Yes — it's negligible (sync) |
| Network DB (Postgres/MySQL) | 1–100ms+ | ❌ No — must be async (real I/O wait) |

bcrypt is CPU-bound *and slow* → offload it. SQLite is fast *and local* → the async machinery is pure overhead. A network database is the classic async case because the thread genuinely sits idle waiting on the wire.

**TLDR:** Async isn't "good," sync isn't "bad." Async only pays off when the thing you're offloading is either slow (bcrypt) or a genuine wait on external I/O (network). For an embedded store that returns in microseconds, sync is faster *because* you skip the coordination cost. Same principle, opposite conclusion.
